### PR TITLE
datatype/romio: fixing Type_create_darray plus romio flatten

### PR DESCRIPTION
**Pull request by markalle**
Two fixes here:
1. darray needs a resize in the block()/cyclic() subtype creation
2. romio flatten has flat->count consistency problems

When Type_create_darray() uses the block()/cyclic() function to
construct a subtype for whatever dimension it's processing, it
needs to resize the resulting type before it goes into the type
construction as the input for processing the next dimension.

(The same problem is in the main path's type creation, and in romio's
mirroring of it).

Gist for a testcase:
https://gist.github.com/markalle/940de93d64fd779e304ee124855b8e6a

The darray_bug_romio.c testcase creates a darray using
* 4 ranks in a 2x2 grid
* looking at the type made for rank 0
* inner dimension: 4 ints distributed block over 2 ranks with 2 items each
* outer dimension: 6 of the above distributed cyclic over 2 ranks with 2 items each

The type created by processing the first dimension should look like this
[ x x . . ]

And then distributing 6 of the above for the outer dimension would be
[ x x . . ]
[ x x . . ]
[ . . . . ]
[ . . . . ]
[ x x . . ]
[ x x . . ]

But if we remove this PR's resize at the bottom of MPIOI_Type_block()
for example, the first dimension is just
[ x x ]

And then distributing those for the second dimension becomes

[ x x x x ]
[ . . . . ]
[ . . . . ]
[ . . . . ]
[ x x x x ]
[ . . . . ]

Going to the MPI standard to justify why the first layout is right,
it's where the definiton of the cyclic() function has a ub_marker
of gsize*ex, eg 4 ints for that first type.

I did notice datatype_impl.c has a resize already in its cyclic() function,
but I wasn't sure if that was trying to handle this case or not, and that
resize isn't mirrored in the romio code.  Either way it's not handling the
case right as "ex" in the standard isn't the orig_extent, it's the extent
of oldtype that came from the previous dimension.  Eg in the outer dimension
above gsize=6 and ex=16 when setting the size for the final type.

--------
Second change, orthogonal to the above, for romio Flatten:

ADIOI_Count_contiguous_blocks and ADIOI_Flatten are very similar functions,
but they diverge a lot around the edges and in degenerate cases.  In Spectrum
MPI I spent some time making them consistent with each other but
found that to be a losing battle.

So the approach I used here is to not have Count() be as definitive,
and rather let Flatten() have the last word on what the final flat->count
really is.  Eg Flatten's *curr_index is the real count.

The changes I made are

1. Fix a couple places in Flatten where *curr_index was updated out
of sync with what was actually being added to flat->indices[] and
flat->blocklens[].  That's one of the important book-keeping rules
Flatten should follow.  There were a couple places (when counts[] arrays
have 0s) where that wasn't the case (see the "nonzeroth" var in this
commit).

2. The main change I made was to reset flat->count based on
Flatten's curr_index.  This is because the divergence between the
two functions is too great to reliably fix.

3. A third change is just a safety valve, using a
flatlist_node_grow() function just in case the
Count function returns a smaller value than Flatten ends up
trying to write into the array, and related to this I
got rid of the various --(flat->count) lines, since that var now
represents the allocated size of the array, until right after
the Flatten function when it's reset to represent the data written
to the array like it used to be.

I don't think we even need ADIOI_Count_contiguous_blocks() anymore,
but I didn't remove it.

Signed-off-by: Mark Allen <markalle@us.ibm.com>

***************************************************************
I'm not sure if IBM has a signed contributors license yet, but I figure
we can do that in the background while this PR is being reviewed.

### CODE REVIEW

File: src/mpi/datatype/datatype_impl.c
Comment (hzhou) [inline src/mpi/datatype/datatype_impl.c:3]: External copyright contributions is noted in the top level COPYRIGHT file. We no longer note additional copyright in individual files.

Comment (markalle) [reply src/mpi/datatype/datatype_impl.c]: Thanks, I just now removed them

**Comment by markalle** [comment]
Repushed: removed copyright notices from the individual files

**Comment by hzhou** [comment]
test:mpich/authorship

**Comment by hzhou** [comment]
@markalle Could you send us an updated contribution agreement per https://github.com/pmodels/mpich/blob/main/CONTRIBUTING . We'll need your name explicitly included in the designated employee list. 

Secondly, could you split the commit into possibly 3 commits, one for `datatype_impl.c`, one for Romio's datatype code, and one for the Romio flatten behavior change?

**Comment by hzhou** [comment]
@roblatham00 Could you review the ROMIO flatten count commit?

**Comment by markalle** [comment]
Thanks, I split the PR into 3 commits.

About the contributor agreement, am I understanding right that the corporate license agreements involve a centralized list that I need to locate the maintainer of and get that updated, as opposed to me just signing a document on an individual basis as myself?

**Review by hzhou — APPROVED** [review]
The datatype changes looks reasonable to me.

Approved pending on tests and @roblatham00 's review of the flatten commit.

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most
test:mpich/warnings

**Comment by hzhou** [comment]
> Thanks, I split the PR into 3 commits.
> 
> About the contributor agreement, am I understanding right that the corporate license agreements involve a centralized list that I need to locate the maintainer of and get that updated, as opposed to me just signing a document on an individual basis as myself?

I think if you want to (and can) contribute on an individual basis, you may choose to sign on an individual basis. You probably want to check with your legal team to be sure.

**Comment by roblatham00** [comment]
bravo to you for digging into the darray code.   RESIZED came to the standard after the darray code was written, but here in 2020 I can't imagine there are any MPI-1 implementations out there -- not even MS-MPI!

It's always bothered me that we made one pass over the type to count and allocate the blocks, then a second pass to do the work (and a third pass to coalesce any adjacent blocks).   So far it hasn't been a noticeable overhead but I still appreciate removing the call to count_contiguous blocks.

I tested this against the romio test suite (src/mpi/romio/tests) which has one test specifically to catch the "types with zeros" case.   I appreciate the second set of eyes on that area of the code.  Glad to report that all the C tests pass. (I did not try the fortran tests: having a momemtary build problem with the fortran bindings. 

Can we add `darray_bug_romio.c` to  either the  tests in src/mpi/romio/tests or the tests in `test/mpi/io` ?

**Comment by hzhou** [comment]
> Can we add `darray_bug_romio.c` to either the tests in src/mpi/romio/tests or the tests in `test/mpi/io` ?

Sure. And I think we *should* add any missing romio tests to `test/mpi/io`.

**Review by roblatham00 — APPROVED** [review]

**Comment by hzhou** [comment]
test:mpich/authorship

**Comment by hzhou** [comment]
@markalle It is ready to be merged pending contribution agreement.

**Comment by markalle** [comment]
Thanks again, I just e-mailed in my signed contributor agreement PDF

**Comment by hzhou** [comment]
test:mpich/warnings

**Pull request merged**
